### PR TITLE
feat(config,cache/s3proxy): add s3 addressing style support

### PR DESCRIPTION
**Pull request by wuhuizuo**
Some S3 providers do not support auto bucket lookup type, need to set it with `path` type or `dns` type. 

Here I added a support to set it in `s3_proxy` filed in configuration to make it reachable.

### CODE REVIEW

File: cache/s3proxy/s3proxy.go [hunk 1]
```diff
@@ -49,6 +49,7 @@ func New(
 	// S3CloudStorageConfig struct fields:
 	Endpoint string,
 	Bucket string,
+	BucketLookupType minio.BucketLookupType,
```
Comment (mostynb) [inline cache/s3proxy/s3proxy.go:52]: Could you add a command line flag for this too, along with the other `--s3.*` options?

**Review by mostynb — COMMENTED** [review]
Thanks for the contribution. Sorry it took me a few days to look at this.

File: cache/s3proxy/s3proxy.go (hunk 1 above)
Comment (wuhuizuo) [reply cache/s3proxy/s3proxy.go:52]: Added, PTAL.

**Comment by wuhuizuo** [comment]
> Thanks for the contribution. Sorry it took me a few days to look at this.

I fixed it.

File: utils/flags/flags.go [hunk 2]
```diff
@@ -219,6 +219,12 @@ func GetCliFlags() []cli.Flag {
 			Usage:   "The S3/minio bucket to use when using S3 proxy backend.",
 			EnvVars: []string{"BAZEL_REMOTE_S3_BUCKET"},
 		},
+		&cli.IntFlag{
```
Comment (mostynb) [inline utils/flags/flags.go:222]: Most of bazel-remotes other flags take string values, would you mind changing this to do that? We could use the values "dns", "path" and "auto". I think that's a bit easier to understand when reading a config file or command line arguments.

**Review by mostynb — CHANGES_REQUESTED** [review]
This feature looks good to me, but I think we should use string flags instead, for improved readability.

I had to do some background reading to understand the use case. This blog post was useful, sharing it here for anyone reading through this PR later: https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/

File: utils/flags/flags.go (hunk 2 above)
Comment (wuhuizuo) [reply utils/flags/flags.go]: Yes, your suggestion fine. I updated it.

File: config/proxy.go [hunk 3]
```diff
@@ -78,3 +80,14 @@ func (c *Config) setProxy() error {
...
+
+func parseBucketLookupType(typeStr string) minio.BucketLookupType {
+	valMap := map[string]minio.BucketLookupType{
+		"auto": minio.BucketLookupAuto,
+		"dns":  minio.BucketLookupDNS,
+		"path": minio.BucketLookupPath,
+	}
+
+	// also when not found the type, return "auto" type.
+	return valMap[typeStr]
```
Comment (mostynb) [inline config/proxy.go:92]: I think this function should probably check if `typeStr` is in the map, and return an error if it isn't. And this error checking should be done in `validateConfig()` in config/config.go

File: utils/flags/flags.go [hunk 4]
```diff
@@ -219,6 +219,12 @@ func GetCliFlags() []cli.Flag {
 			Usage:   "The S3/minio bucket to use when using S3 proxy backend.",
 			EnvVars: []string{"BAZEL_REMOTE_S3_BUCKET"},
 		},
+		&cli.StringFlag{
+			Name:    "s3.bucket_lookup_type",
+			Value:   "auto",
+			Usage:   "The S3/minio bucket lookup type(auto|dns|path) to use when using S3 proxy backend.",
```
Comment (mostynb) [inline utils/flags/flags.go:225]: nitpick: could you reword this to something like: `The S3/minio bucket lookup type. Allowed values: auto, dns, path.` and update the help text in README.md to match?

File: README.md [hunk 5]
```diff
@@ -263,6 +263,10 @@ OPTIONS:
    --s3.bucket value The S3/minio bucket to use when using S3 proxy backend.
       [$BAZEL_REMOTE_S3_BUCKET]
 
+   --s3.bucket_lookup_type value The S3/minio bucket lookup type to use when
```
Comment (mostynb) [inline README.md:266]: github won't let me comment on lines that aren't close to the dirr, but could you also add bucket_lookup_type to the example s3_proxy yaml further down in this file?

File: config/proxy.go (hunk 3 above)
Comment (wuhuizuo) [reply config/proxy.go:92]: I think we only need to check it in `validateConfig()`.

Comment (wuhuizuo) [reply config/proxy.go]: Also I refactored it again.

**Comment by mostynb** [comment]
Thanks for following through with these changes- I made a few minor tweaks and landed this on the master branch.
